File: packages/client-swift/Sources/SpriteCoreClient/SpriteAnimationPlayer.swift
import Foundation

/// Platform-independent playback engine. One instance per character per mode.
/// Drives `currentRef` forward over time according to the `AnimationGraph`'s
/// animations and transitions; callers materialize frames via their own
/// `FrameSource`.
///
/// Mirrors the Kotlin `SpriteAnimationPlayer` and the TS `SpriteAnimationPlayer`.
/// Observability is via `AsyncStream<FrameRef?>` and `AsyncStream<String>` — a
/// SwiftUI consumer can also read the latest synchronous values through
/// `currentRef` / `currentState`.
public actor SpriteAnimationPlayer {
    private let graph: AnimationGraph
    private let ticker: Ticker
    private let minFrameDelayMs = 16

    private var _currentRef: FrameRef?
    private var _currentState: String
    private var refContinuations: [UUID: AsyncStream<FrameRef?>.Continuation] = [:]
    private var stateContinuations: [UUID: AsyncStream<String>.Continuation] = [:]

    private var runningTask: Task<Void, Never>?

    public init(graph: AnimationGraph, ticker: Ticker = SystemTicker()) {
        self.graph = graph
        self.ticker = ticker
        self._currentState = graph.defaultState
        // Kick off default-state playback on actor init. Use a detached Task
        // because the actor isn't fully initialized until this init returns.
        Task { [weak self] in
            await self?.startDefaultState()
        }
    }

    // MARK: - Public surface

    public var currentRef: FrameRef? { _currentRef }
    public var currentState: String { _currentState }

    public func refStream() -> AsyncStream<FrameRef?> {
        AsyncStream { continuation in
            let id = UUID()
            continuation.yield(_currentRef)
            refContinuations[id] = continuation
            continuation.onTermination = { [weak self] _ in
                Task { await self?.removeRefContinuation(id) }
            }
        }
    }

    public func stateStream() -> AsyncStream<String> {
        AsyncStream { continuation in
            let id = UUID()
            continuation.yield(_currentState)
            stateContinuations[id] = continuation
            continuation.onTermination = { [weak self] _ in
                Task { await self?.removeStateContinuation(id) }
            }
        }
    }

    /// Request a state change. If the graph's transitions table has a match
    /// for `currentState → target`, that transition plays once before the
    /// target state's own loop starts.
    ///
    /// `playCount` semantics (from `<<<state-N>>>`):
    ///   - nil or 0 — loop indefinitely
    ///   - N >= 1   — play N times and hold last frame
    public func requestState(_ target: String, playCount: Int? = nil) async {
        let previousState = _currentState
        let sameState = target == previousState
        let effectiveCount: Int? = (playCount ?? 0) >= 1 ? playCount : nil

        await cancelRunning()

        if sameState && effectiveCount == nil {
            return
        }

        let task = Task { [weak self] in
            guard let self else { return }
            if !sameState {
                if case let .phase(ref) = await self.graph.resolveTransition(from: previousState, to: target) ?? .phase("") {
                    if !ref.isEmpty {
                        let resolved = ResolvedTransition.parse(ref)
                        await self.playPhase(animName: resolved.animation, phase: resolved.phase, loopOverride: .once)
                    }
                }
            }
            await self.playState(target, entering: !sameState, playCountOverride: effectiveCount)
        }
        runningTask = task
    }

    public func dispose() async {
        await cancelRunning()
        for (_, c) in refContinuations { c.finish() }
        for (_, c) in stateContinuations { c.finish() }
        refContinuations.removeAll()
        stateContinuations.removeAll()
    }

    // MARK: - Internals

    private func removeRefContinuation(_ id: UUID) {
        refContinuations.removeValue(forKey: id)
    }

    private func removeStateContinuation(_ id: UUID) {
        stateContinuations.removeValue(forKey: id)
    }

    private func startDefaultState() async {
        let task = Task { [weak self] in
            guard let self else { return }
            await self.playState(await self.graph.defaultState, entering: true, playCountOverride: nil)
        }
        runningTask = task
    }

    private func cancelRunning() async {
        runningTask?.cancel()
        runningTask = nil
    }

    private func setRef(_ ref: FrameRef?) {
        _currentRef = ref
        for (_, c) in refContinuations { c.yield(ref) }
    }

    private func setState(_ state: String) {
        if _currentState != state {
            _currentState = state
            for (_, c) in stateContinuations { c.yield(state) }
        }
    }

    private func playState(_ state: String, entering: Bool, playCountOverride: Int?) async {
        setState(state)
        guard let anim = graph.animations[state] else { return }
        if entering, anim.intro != nil {
            await playPhase(animName: state, phase: .intro, loopOverride: nil)
            if Task.isCancelled { return }
        }
        if let count = playCountOverride, count >= 1 {
            await playPhaseFinite(animName: state, phase: .loop, times: count)
            return
        }
        await playPhase(animName: state, phase: .loop, loopOverride: nil)
    }

    private func playPhaseFinite(animName: String, phase: Phase, times: Int) async {
        guard let anim = graph.animations[animName], let seq = pickPhase(anim, phase), !seq.frames.isEmpty else { return }
        let frameDelayMs = max(1000 / seq.fps, minFrameDelayMs)
        for _ in 0..<times {
            for ref in seq.frames {
                if Task.isCancelled { return }
                setRef(ref)
                do { try await ticker.delay(ms: frameDelayMs) } catch { return }
            }
        }
        if let last = seq.frames.last { setRef(last) }
        // Hold indefinitely until cancelled.
        while !Task.isCancelled {
            do { try await ticker.delay(ms: 1000) } catch { return }
        }
    }

    private func playPhase(animName: String, phase: Phase, loopOverride: LoopMode?) async {
        guard let anim = graph.animations[animName], let seq = pickPhase(anim, phase), !seq.frames.isEmpty else { return }
        let frameDelayMs = max(1000 / seq.fps, minFrameDelayMs)
        let loop = loopOverride ?? seq.loop

        switch loop {
        case .once:
            for ref in seq.frames {
                if Task.isCancelled { return }
                setRef(ref)
                do { try await ticker.delay(ms: frameDelayMs) } catch { return }
            }
            if !seq.holdLastFrame { setRef(nil) }
        case .pingPong:
            let cap = seq.iterations ?? .max
            var rounds = 0
            while rounds < cap, !Task.isCancelled {
                for ref in seq.frames {
                    if Task.isCancelled { return }
                    setRef(ref)
                    do { try await ticker.delay(ms: frameDelayMs) } catch { return }
                }
                if seq.frames.count > 2 {
                    for i in stride(from: seq.frames.count - 2, through: 1, by: -1) {
                        if Task.isCancelled { return }
                        setRef(seq.frames[i])
                        do { try await ticker.delay(ms: frameDelayMs) } catch { return }
                    }
                }
                rounds += 1
            }
        case .infinite:
            while !Task.isCancelled {
                for ref in seq.frames {
                    if Task.isCancelled { return }
                    setRef(ref)
                    do { try await ticker.delay(ms: frameDelayMs) } catch { return }
                }
            }
        }
    }

    private func pickPhase(_ anim: Animation, _ phase: Phase) -> FrameSequence? {
        switch phase {
        case .intro: return anim.intro
        case .outro: return anim.outro
        case .loop:  return anim.effectiveLoop
        }
    }
}

File: packages/client-kotlin/core/src/main/kotlin/ai/openclaw/spritecore/client/SpriteAnimationPlayer.kt
package ai.openclaw.spritecore.client

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.awaitCancellation
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Platform-independent playback engine. One instance per character per mode.
 * Drives [currentRef] forward over time according to the [AnimationGraph]'s
 * animations and transitions; callers materialize frames via their own
 * [FrameSource].
 *
 * Thread safety: [requestState] is safe to call from any thread. Internal
 * state mutations happen on the supplied coroutine scope's dispatcher.
 */
class SpriteAnimationPlayer(
    private val graph: AnimationGraph,
    private val ticker: Ticker = SystemTicker(),
    scope: CoroutineScope? = null,
) {
    private val owned = scope == null
    private val scope: CoroutineScope = scope ?: CoroutineScope(SupervisorJob() + Dispatchers.Default)

    private val _currentRef = MutableStateFlow<FrameRef?>(null)
    /** The frame the caller should be rendering right now. Null = blank. */
    val currentRef: StateFlow<FrameRef?> = _currentRef.asStateFlow()

    private val _currentState = MutableStateFlow(graph.defaultState)
    /** The agent state the player is currently in (post-transition). */
    val currentState: StateFlow<String> = _currentState.asStateFlow()

    private var activeJob: Job? = null

    init {
        activeJob = this.scope.launch {
            playState(graph.defaultState, entering = true)
        }
    }

    /**
     * Request a state change. If the [graph]'s transitions table has a match
     * for `currentState → target`, that transition plays once before the
     * target state's own loop starts.
     *
     * [playCount] semantics (from the client-parsed `<<<state-N>>>` marker):
     *   null or 0 → default: the state's configured loop plays indefinitely
     *               until the next [requestState] cancels it.
     *   N >= 1    → play the loop phase exactly N times, then hold the last
     *               frame indefinitely. Intro (if any) still plays once.
     *
     * When [playCount] is non-null we always replay — even when [target] is
     * already the current state — so a model emitting the same marker twice
     * in a row ("<<<wink-1>>> then <<<wink-1>>>") visibly replays the
     * animation instead of being swallowed as a no-op.
     */
    fun requestState(target: String, playCount: Int? = null): Job {
        // Capture the job reference before reassignment. Reading activeJob
        // from inside the launched block is a race — by the time the block
        // runs, activeJob has been overwritten to `job` itself, so the old
        // job (often the init-spawned infinite loop) would never be cancelled.
        val previousJob = activeJob
        val job = scope.launch {
            val sameState = target == _currentState.value
            if (sameState && (playCount == null || playCount <= 0)) {
                previousJob?.cancelAndJoin()
                return@launch
            }
            val previousState = _currentState.value
            previousJob?.cancelAndJoin()
            if (!sameState) {
                val transition = graph.resolveTransition(previousState, target)
                if (transition is TransitionRef.Phase) {
                    val resolved = ResolvedTransition.parse(transition.value)
                    playPhase(
                        animName = resolved.animation,
                        phase = resolved.phase,
                        loopOverride = LoopMode.ONCE,
                    )
                }
                // Crossfade transitions are currently played as an instant
                // swap; the visual blend is a rendering-side concern the
                // consumer applies when the ref changes.
            }
            playState(target, entering = !sameState, playCountOverride = playCount)
        }
        activeJob = job
        return job
    }

    /** Cancel playback and, if we own it, the internal scope. */
    fun dispose() {
        activeJob?.cancel()
        if (owned) {
            scope.cancel()
        }
    }

    // --- internals ---

    private suspend fun playState(
        state: String,
        entering: Boolean,
        playCountOverride: Int? = null,
    ) {
        _currentState.value = state
        val anim = graph.animations[state] ?: return
        if (entering && anim.intro != null) {
            playPhase(state, Phase.INTRO)
        }
        if (playCountOverride != null && playCountOverride >= 1) {
            playPhaseFinite(state, Phase.LOOP, playCountOverride)
            return
        }
        // Flat states fall through to `effectiveLoop`; phased states play
        // `loop` here. `outro` fires only on requestState() via transitions.
        playPhase(state, Phase.LOOP)
    }

    /**
     * Play the [phase] of [animName] exactly [times] times, then hold the
     * last frame indefinitely (until this coroutine is cancelled by the
     * next [requestState]). Implements the `<<<state-N>>>` N>=1 contract:
     * play N times and pause on the last frame.
     */
    private suspend fun playPhaseFinite(
        animName: String,
        phase: Phase,
        times: Int,
    ) {
        val anim = graph.animations[animName] ?: return
        val seq = when (phase) {
            Phase.INTRO -> anim.intro
            Phase.LOOP -> anim.effectiveLoop
            Phase.OUTRO -> anim.outro
        } ?: return
        if (seq.frames.isEmpty()) return
        val frameDelayMs = (1000L / seq.fps).coerceAtLeast(MIN_FRAME_DELAY_MS)
        repeat(times) {
            for (ref in seq.frames) {
                _currentRef.value = ref
                ticker.delay(frameDelayMs)
            }
        }
        _currentRef.value = seq.frames.last()
        awaitCancellation()
    }

    private suspend fun playPhase(
        animName: String,
        phase: Phase,
        loopOverride: LoopMode? = null,
    ) {
        val anim = graph.animations[animName] ?: return
        val seq = when (phase) {
            Phase.INTRO -> anim.intro
            Phase.LOOP -> anim.effectiveLoop
            Phase.OUTRO -> anim.outro
        } ?: return
        if (seq.frames.isEmpty()) {
            return
        }
        val frameDelayMs = (1000L / seq.fps).coerceAtLeast(MIN_FRAME_DELAY_MS)
        val loop = loopOverride ?: seq.loop

        when (loop) {
            LoopMode.ONCE -> {
                for (ref in seq.frames) {
                    _currentRef.value = ref
                    ticker.delay(frameDelayMs)
                }
                if (!seq.holdLastFrame) {
                    _currentRef.value = null
                }
            }
            LoopMode.PING_PONG -> {
                val cap = seq.iterations ?: Int.MAX_VALUE
                var rounds = 0
                while (rounds < cap) {
                    for (ref in seq.frames) {
                        _currentRef.value = ref
                        ticker.delay(frameDelayMs)
                    }
                    for (i in seq.frames.size - 2 downTo 1) {
                        _currentRef.value = seq.frames[i]
                        ticker.delay(frameDelayMs)
                    }
                    rounds++
                }
            }
            LoopMode.INFINITE -> {
                while (true) {
                    for (ref in seq.frames) {
                        _currentRef.value = ref
                        ticker.delay(frameDelayMs)
                    }
                }
            }
        }
    }

    private suspend fun Job.cancelAndJoin() {
        cancel()
        try {
            join()
        } catch (_: Throwable) {
            // Cancellation unwinds through here; swallow so caller's flow continues.
        }
    }

    companion object {
        private const val MIN_FRAME_DELAY_MS = 16L
    }
}
